op-chain-ops/cmd/op-run-block/main.go

```diff
@@ -0,0 +1,363 @@
+		r.logger.Error("fetched incompatible header", "expectedHash", hash, "fetchedHash", got, "num", u)
+	}
+	return hdr
+}
+
+func Process(logger log.Logger, config *params.ChainConfig,
+	block *sources.RPCBlock,
+	statedb *gstate.StateDB, cfg vm.Config,
+	chainCtx *remoteChainCtx, outW io.Writer) (*core.ProcessResult, error) {
+	var (
+		receipts    types.Receipts
+		usedGas     = new(uint64)
+		header      = block.CreateGethHeader()
+		blockHash   = block.Hash
+		blockNumber = new(big.Int).SetUint64(uint64(block.Number))
+		allLogs     []*types.Log
+		gp          = new(core.GasPool).AddGas(uint64(block.GasLimit))
+	)
+
+	// Mutate the block and state according to any hard-fork specs
+	if config.DAOForkSupport && config.DAOForkBlock != nil && config.DAOForkBlock.Cmp(blockNumber) == 0 {
+		misc.ApplyDAOHardFork(statedb)
+	}
+	misc.EnsureCreate2Deployer(config, uint64(block.Time), statedb)
+	var (
+		context vm.BlockContext
+		signer  = types.MakeSigner(config, header.Number, header.Time)
+		err     error
+	)
+	context = core.NewEVMBlockContext(header, chainCtx, nil, config, statedb)
+	vmenv := vm.NewEVM(context, vm.TxContext{}, statedb, config, cfg)
+	if beaconRoot := block.ParentBeaconRoot; beaconRoot != nil {
+		core.ProcessBeaconBlockRoot(*beaconRoot, vmenv, statedb)
+	}
+	if config.IsPrague(blockNumber, uint64(block.Time)) {
+		core.ProcessParentBlockHash(block.ParentHash, vmenv, statedb)
+	}
+	logger.Info("Prepared EVM state")
+	_, _ = fmt.Fprintf(outW, "# Prepared state\n")
+
+	// Iterate over and process the individual transactions
+	for i, tx := range block.Transactions {
+		logger.Info("Processing tx", "i", i, "hash", tx.Hash())
+		_, _ = fmt.Fprintf(outW, "# Processing tx %d\n", i)
+		msg, err := core.TransactionToMessage(tx, signer, header.BaseFee)
+		if err != nil {
+			return nil, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
+		}
+		statedb.SetTxContext(tx.Hash(), i)
+
+		receipt, err := core.ApplyTransactionWithEVM(msg, config, gp, statedb, blockNumber, blockHash, tx, usedGas, vmenv)
+		if err != nil {
+			return nil, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
+		}
+		receipts = append(receipts, receipt)
+		allLogs = append(allLogs, receipt.Logs...)
+	}
+	logger.Info("Done with transactions")
+	_, _ = fmt.Fprintf(outW, "# Done with transactions\n")
+
+	// Read requests if Prague is enabled.
+	var requests types.Requests
+	if config.IsPrague(new(big.Int).SetUint64(uint64(block.Number)), uint64(block.Time)) {
+		requests, err = core.ParseDepositLogs(allLogs, config)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	engine := chainCtx.Engine()
+	// Finalize the block, applying any consensus engine specific extras (e.g. block rewards)
+	engine.Finalize(chainCtx, header, statedb,
+		&types.Body{Transactions: block.Transactions, Withdrawals: *block.Withdrawals})
+	logger.Info("Completed block processing")
+	_, _ = fmt.Fprintf(outW, "# Completed block processing\n")
```

Comment:
Why must the block become finalized?

Reply:
It's geth-specific terminology for sealing the block.